File: Inheritance.py
In Ruby, inheritance allows you to create a new class based on an existing class,
inheriting its attributes and behaviors. This promotes code reuse and allows you to create specialized classes 
that inherit and extend the functionality of their parent classes. 
Heres an explanation of inheritance and method overriding in Ruby with code samples:

Inheritance:

To create a subclass that inherits from a superclass, you use the < symbol followed by the name of the superclass.
Example:

class Animal
  def speak
    puts "Animal speaks."
  end
end

class Dog < Animal
  def bark
    puts "Dog barks."
  end
end





Method Inheritance:

Subclasses inherit all the methods defined in their superclass.
They can use those methods directly or override them with their own implementations.

Example:

class Animal
  def speak
    puts "Animal speaks."
  end
end

class Dog < Animal
  def bark
    puts "Dog barks."
  end
end

animal = Animal.new
animal.speak  # Output: "Animal speaks."

dog = Dog.new
dog.speak     # Output: "Animal speaks."
dog.bark      # Output: "Dog barks."





Method Overriding:

Method overriding allows a subclass to provide its own implementation of a method inherited from the superclass. 
This allows for customization and specialization.
Example:

class Animal
  def speak
    puts "Animal speaks."
  end
end

class Dog < Animal
  def speak
    puts "Dog barks."
  end
end

animal = Animal.new
animal.speak  # Output: "Animal speaks."

dog = Dog.new
dog.speak     # Output: "Dog barks."

Inheritance and method overriding enable you to build hierarchies of classes,
where subclasses inherit and extend the behavior of their parent classes. 
This helps in organizing code, promoting code reuse, and achieving polymorphism.
Subclasses can have their own unique behaviors while inheriting and reusing common functionalities defined in their superclasses.









MORE EXPLANTION
********************************************************************************************************************************



Object methods in Ruby are methods that are defined within a class but are not directly associated with an object.
Instead, they are associated with the class itself. 
These methods can be used to perform operations on the class or to maintain state information for the class geeksforgeeks.org.

Here's an example using a Person class with a class method:

class Person
  def initialize(name, age, sex)
    @name = name
    @age = age
    @sex = sex
  end

  def display_person
    puts "The name of the person is #{@name} and age is #{@age} also he is a #{@sex}"
  end

  def self.greater_than(other_person, attribute)
    @age > other_person.instance_variable_get(:@age)
  end
end

person1 = Person.new("John", 30, "male")
person2 = Person.new("Jane", 25, "female")

puts person1.greater_than(person2, :age) # => true
In this example, we define a class method greater_than within the Person class.
This method compares the age of self (the calling object) with the age of the other_person object passed as an argument. 
The instance_variable_get method is used to access the @age instance variable of the other_person object stackoverflow.com.

For more information and examples, you can refer to the provided sources










MORE EXPLANTION
********************************************************************************************************************************



Sure, object methods in Ruby are functions that are defined within a class and can be called on objects of that class.
They are used to perform actions on the object, such as changing its state or returning information about it.

Here is an example of an object method:

Ruby
class Person
  def initialize(name)
    @name = name
  end

  def say_hello
    puts "Hello, #{name}"
  end
end
Use code with caution. Learn more
In this example, the say_hello method is defined within the Person class. 
It takes no arguments and returns nothing. When an object of the Person class is created,
the say_hello method can be called on it. For example:

Ruby
person = Person.new("John Doe")
person.say_hello # => Hello, John Doe
Use code with caution. Learn more
The say_hello method is an example of an instance method. 
Instance methods are methods that are defined within a class and can only be called on objects of that class.

There are also class methods, which are methods that are defined within a class and can be called on the class itself.
Class methods are prefixed with the self keyword. For example:

Ruby
class Person
  def self.all
    # This method returns an array of all Person objects
  end
end

Person.all # => [Person(name: "John Doe"), Person(name: "Jane Doe")]
Use code with caution. Learn more
Object methods are a powerful way to add functionality to your classes.
They can be used to perform actions on objects, change their state, or return information about them.

Here are some additional things to note about object methods:

Object methods are defined within a class.
Object methods can take arguments and return values.
Object methods can be called on objects of the class they are defined in.
Object methods can be called on the class itself.
I hope this helps!
































































































































































































..p

File: Classes/Inheritance.java
Inheritance in Ruby is a fundamental object-oriented programming concept that allows one class to inherit the properties and behaviors (methods and attributes) of another class. It enables code reuse, extensibility, and the creation of class hierarchies.

Here's how inheritance works in Ruby:

1. **Superclass and Subclass**:
   - Inheritance involves two classes: a superclass (parent class) and a subclass (child class).
   - The superclass is the class from which you want to inherit properties and behaviors.
   - The subclass is the new class that inherits from the superclass.

2. **Defining a Superclass**:
   - You define a superclass by creating a class with the methods and attributes you want to share.
   - For example:

   ```ruby
   class Animal
     def speak
       puts "Animal speaks"
     end
   end
   ```

3. **Defining a Subclass**:
   - To create a subclass, use the `<` symbol followed by the name of the superclass in the class definition.
   - The subclass inherits all the methods and attributes of the superclass.

   ```ruby
   class Dog < Animal
     def bark
       puts "Woof! Woof!"
     end
   end
   ```

4. **Inheriting Methods**:
   - The subclass inherits all methods from the superclass. In this example, `Dog` inherits the `speak` method from `Animal`.
   - You can also add new methods to the subclass.

5. **Overriding Methods**:
   - Subclasses can override methods from the superclass by redefining them in the subclass.
   - This allows you to provide custom implementations for methods in the subclass.

   ```ruby
   class Dog < Animal
     def speak
       puts "Dog barks"
     end
   end
   ```

6. **Calling Superclass Methods**:
   - In overridden methods, you can call the superclass's implementation using the `super` keyword.
   - This allows you to extend the behavior of the superclass's method in the subclass.

   ```ruby
   class Dog < Animal
     def speak
       super
       puts "Dog barks"
     end
   end
   ```

7. **Using `is_a?` and `kind_of?`**:
   - You can check if an object is an instance of a particular class or its subclasses using the `is_a?` or `kind_of?` methods.

   ```ruby
   my_dog = Dog.new
   puts my_dog.is_a?(Animal) # true
   puts my_dog.is_a?(Dog)    # true
   ```

8. **Using `ancestors`**:
   - The `ancestors` method allows you to see the inheritance hierarchy of a class.
   - It returns an array of classes in the order they are checked when a method is called on an object.

   ```ruby
   puts Dog.ancestors
   # [Dog, Animal, Object, Kernel, BasicObject]
   ```

Inheritance in Ruby is a powerful way to create class hierarchies, promote code reuse, and achieve a more organized and maintainable codebase. Subclasses can add specific functionality while inheriting common behavior from the superclass.





  ......
